# SQL - Анализ базы данных

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Моя первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

**В проекте будут использованы следующие библиотеки:**

In [1]:
import pandas as pd
from sqlalchemy import create_engine 
from sqlalchemy import create_engine, text as sql_text
from IPython.display import Image

**Ниже представлена схема базы данных:**

In [2]:
# get the image
Image(url="https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&table=block&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=1640&userId=&cache=v2", width=900, height=900)

### Задания

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайть среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

### Ход выполнения заданий

- Описать цели исследования;
- Исследовать таблицы — вывести первые строки;
- Сделайть по одному SQL-запросу для решения каждого задания;
- Вывести результат каждого запроса в тетрадке;
- Описать выводы по каждой из решённых задач.

### Доступ к базе данных

In [3]:
# Устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
connection = create_engine(connection_string)
# Сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Примечания

- Не забываем про функции, они значительно облегчат жизнь при выполнении запросов;
- Получием ответы, используя только язык запросов SQL. 

Напишу функцию для выполнения запроса:

In [4]:
def request(query):
    return pd.io.sql.read_sql(query, con = engine)

**books:**

In [5]:
books = '''SELECT * FROM books'''
request(books).head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**authors:**

In [6]:
authors = '''SELECT * FROM authors'''
request(authors).head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**publishers:**

In [7]:
publishers = '''SELECT * FROM publishers'''
request(publishers).head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**ratings:**

In [8]:
ratings = '''SELECT * FROM ratings'''
request(ratings).head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**reviews:**

In [9]:
reviews = '''SELECT * FROM reviews'''
request(reviews).head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**Посчитаем, сколько книг вышло после 1 января 2000 года**

In [10]:
first_task = '''SELECT COUNT(DISTINCT book_id) FROM books
             WHERE publication_date >= '2000-01-01';'''
request(first_task)

count
0    821

Ответ: 821 книга вышло после 1 января 2000 года включительно.

**Вывод:** 

* Посчитано количество книг вышедших после 1 января 2000 года включительно

**Для каждой книги посчитаем количество обзоров и среднюю оценку**

In [11]:
second_task = '''SELECT title,
                              books.book_id,
                              COUNT(DISTINCT reviews.review_id) AS number_of_reviews,
                              AVG(ratings.rating) AS average_rating
                       FROM books
                       LEFT JOIN reviews ON reviews.book_id = books.book_id
                       LEFT JOIN ratings ON ratings.book_id = books.book_id
                       GROUP BY books.book_id, books.title
                       ORDER BY average_rating DESC
                       LIMIT 5
                    '''
request(second_task)

title  book_id  number_of_reviews  \
0  A Woman of Substance (Emma Harte Saga #1)       55                  2   
1            Act of Treason (Mitch Rapp  #9)       57                  2   
2              A Dirty Job (Grim Reaper  #1)       17                  4   
3      A Fistful of Charms (The Hollows  #4)       20                  2   
4                              Alas  Babylon       62                  2   

   average_rating  
0             5.0  
1             5.0  
2             5.0  
3             5.0  
4             5.0

**Вывод:** 

* Для каждой книги посчитаны количество обзоров и средняя оценка
* На основе книг с высокими оценками и с высоким количеством обзоров можно сформировать базу литературы для нового приложения

**Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [12]:
third_task = '''SELECT publishers.publisher,
                             publishers.publisher_id,
                             COUNT(DISTINCT books.book_id) AS number_of_books
                       FROM publishers
                       LEFT JOIN books ON publishers.publisher_id = books.publisher_id
                       WHERE books.num_pages > 50
                       GROUP BY publishers.publisher,
                                publishers.publisher_id
                       ORDER BY number_of_books DESC
                       LIMIT 1
                    '''
request(third_task)

publisher  publisher_id  number_of_books
0  Penguin Books           212               42

**Вывод:** 

* Выявлено издательство, выпустившее наибольшее число книг, не включая брошюры
* С издательством Penguin Books можно наладить отношения для успешного продвижения мобильного приложения

**Определим автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками**

In [13]:
four_task = '''SELECT authors.author,
                            AVG(ratings.rating) AS avg_rating,
                            COUNT(rating_id) AS rating_count
                       FROM authors
                       LEFT JOIN books ON authors.author_id = books.author_id
                       LEFT JOIN ratings ON books.book_id = ratings.book_id
                       WHERE ratings.book_id IN (SELECT book_id
                                                 FROM ratings
                                                 GROUP BY book_id
                                                 HAVING COUNT(rating_id) >= 50)
                       GROUP BY authors.author
                       ORDER BY avg_rating DESC
                       LIMIT 1
                    '''
request(four_task)

author  avg_rating  rating_count
0  J.K. Rowling/Mary GrandPré    4.287097           310

**Вывод:** 

* Вычислены авторы с самой высокой средней оценкой книг (учитены только книги с 50 и более оценками)
* С авторами J.K. Rowling/Mary GrandPré можно наладить отношения для успешного продвижения мобильного приложения

**Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок**

In [14]:
five_task = '''WITH t AS 
    (SELECT username,
            COUNT(reviews.review_id) AS reviews_count
     FROM reviews
     WHERE username IN (SELECT username
                        FROM ratings
                        GROUP BY username
                        HAVING COUNT(rating) > 48)
     
     GROUP BY username)
     
SELECT ROUND(AVG(t.reviews_count)) AS avg_reviews
FROM t
'''

request(five_task)

avg_reviews
0         24.0

**Вывод:** 

* Пользователи, которые любят оставлять оценки, в среднем пишут около 24-х обзоров.

**Опишем выводы по каждой из решённых задач**

* Посчитано количество книг вышедших после 1 января 2000 года включительно
* Для каждой книги посчитаны количество обзоров и средняя оценка
* На основе книг с высокими оценками и с высоким количеством обзоров можно сформировать базу литературы для нового приложения
* Выявлено издательство, выпустившее наибольшее число книг, не включая брошюры
* С издательством Penguin Books можно наладить отношения для успешного продвижения мобильного приложения
* Вычислены авторы с самой высокой средней оценкой книг (учитены только книги с 50 и более оценками)
* С авторами J.K. Rowling/Mary GrandPré можно наладить отношения для успешного продвижения мобильного приложения
* Пользователи, которые любят оставлять оценки, в среднем пишут около 24-х обзоров.

На основе выводов сделанных в этой исследовательской работы можно сформировать оптимальную стратеги продвижения мобильной библиотеки.